src/compiler/checker.ts

```diff
@@ -25306,6 +25307,13 @@ namespace ts {
                         }
                     }
                 }
+
+                if (isPropertyAccessExpression(callExpression.expression)) {
+                    const callAccess = callExpression.expression;
+                    if (isIdentifier(callAccess.name) && callAccess.name.escapedText === "hasOwn" && callExpression.arguments.length === 2 && isStringLiteralLike(callExpression.arguments[1])) {

```

Comment:
A downside to special casing the exact names here is that it would break for cases where the function is assigned to a different name:

```ts
const hasProp = Object.hasOwn;

if (hasProp(o, 'test')) ...
```

I think instead of adding this logic within the compiler itself, we might be able to do it by updating the `.d.ts` definitions that are packaged with tsc. Something like:

```ts
interface ObjectConstructor {
    hasOwn<T, P extends string>(o: T, p: P): o is Extract<T, { [p in P]: unknown }>; 
}
```

Reply:
Agreed, it seems like we have the necessary building blocks to express this already.